File: 13-3-2024_Print matrix in diagonal pattern.cpp

//{ Driver Code Starts

#include<bits/stdc++.h>
using namespace std;


// } Driver Code Ends
/*You are required to complete this method */

class Solution{
  public:
    vector<int> matrixDiagonally(vector<vector<int>>&mat)
    {
         //Your code here
         int n= mat.size();
         
         vector<int> ans;
         
         for(int i=0;i<n;i++){
             if(i%2){
                 for(int j=0;j<=i;j++) ans.push_back(mat[j][i-j]);
             }else{
                 for(int j=i;j>=0;j--) ans.push_back(mat[j][i-j]);
             }
         }
         
         for(int i=1;i<n;i++){
             if((i%2)^(n%2)){
                 for(int j=n-i-1;j>=0;j--) ans.push_back(mat[i+j][n-1-j]);
             }else{
                 for(int j=0;j<n-i;j++) ans.push_back(mat[i+j][n-1-j]);
             }
         }
         
         return ans;
    }
};


//{ Driver Code Starts.
int main()
{
    int T;
    cin>>T;
  
    while(T--)
    {
        int n;
        int k=0;
        //cin>>k;
        cin>>n;
        vector<vector<int>>mat(n, vector<int>(n));
        for(int i = 0; i < n; i++){
            for(int j = 0; j < n; j++){
                cin >> mat[i][j];
            }
        }
        Solution obj;
        vector<int>ans = obj.matrixDiagonally(mat);
        for(auto i: ans)cout << i << " ";
        cout << "\n";
        
       
    }
}
// } Driver Code Ends

File: 13-3-2024_Print matrix in diagonal pattern.py
# Your task is to complete this function

class Solution:
    def matrixDiagonally(self,mat):
        # code here
     ans = []
     row = 0
     col = 0
     n = len(mat)
     while len(ans) < n * n:
      while row >= 0 and col < n:
        if row >= 0 and col >= 0 and row < n and col < n:
            ans.append(mat[row][col])
            mat[row][col] = 1e9
        row -= 1
        col += 1
      if len(ans) == n * n:
        break
      row = max(row, 0)
      col = min(col, n - 1)
      while mat[row][col] == 1e9:
        row += 1
      while col >= 0 and row < n:
        if row >= 0 and col >= 0 and row < n and col < n:
            ans.append(mat[row][col])
            mat[row][col] = 1e9
        row += 1
        col -= 1
      if len(ans) == n * n:
         break
      row = min(row, n - 1)
      col = max(col, 0)
      while mat[row][col] == 1e9:
        col += 1
     return ans



#{ 
 # Driver Code Starts
# Driver Program
if __name__=='__main__':
    t = int(input())
    for i in range(t):
        n = list(map(int, input().strip().split()))
        arr = list(map(int, input().strip().split()))
        matrix = [[0 for i in range(n[0])]for j in range(n[0])]
        k=0
        for i in range(n[0]):
            for j in range(n[0]):
                matrix[i][j] = arr[k]
                k+=1
        a = Solution().matrixDiagonally(matrix)
        for x in a:
            print(x,end=' ')
        print('')
# Contributed by: Harshit Sidhwa
# } Driver Code Ends